File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

 /*
 EXPLANATION
We'll do just normal tree traversal of the given binary tree recursivly.

For finding LCA (lowest common ancestor) we've following conditions for every node in the tree,

But before that, this solutions works under the assumption that both Node 'p' & Node 'q' will present in the tree...

if single one of the node is present in the tree, it'll not work or simply return null.

CONDITIONS: -

if current node is present in same tree as 'p' OR 'q'.

if one of it's subtrees contains 'p' and other 'q' (subtrees means, left_sub_tree and right_sub_tree).

if one of it's subtree contains both 'p' & 'q'.

if none of it's subtrees contains any of 'p' & 'q'.

Note: that's a tricky implementation, but works well under the assumption that 'p' & 'q' will be definitely present.

 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == nullptr || root == p || root == q)
            return root;

        TreeNode* left = lowestCommonAncestor(root->left, p, q);
        TreeNode* right = lowestCommonAncestor(root->right, p, q);

        if(left == nullptr)
            return right;
        else if(right == nullptr)
            return left;
        else
            return root;
    }
};

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */




class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // Base case: If root is null or matches one of p or q, return root
        if (root == null || root == p || root == q) {
            return root;
        }

        // Recursively search for p and q in left and right subtrees
        TreeNode leftLCA = lowestCommonAncestor(root.left, p, q);
        TreeNode rightLCA = lowestCommonAncestor(root.right, p, q);

        // If both left and right subtrees return non-null, root is the LCA
        if (leftLCA != null && rightLCA != null) {
            return root;
        }

        // If only one subtree has a non-null result, return that result
        return (leftLCA != null) ? leftLCA : rightLCA;
    }
}